Add documentation for gateway topology api (#7242)
istio/istio.io: content/en/docs/tasks/traffic-management/ingress/ingress-control/index.md

$ kubectl delete virtualservice httpbin
$ kubectl delete --ignore-not-found=true -f @samples/httpbin/httpbin.yaml@
{{< /text >}}

> ## (Experimental) Configuring network toplogies

Reviewer: ```suggestion
## (Experimental) Configuring network toplogies
```

> Istio provides the ability to control X-Forwarded-For and X-Forwarded-Client-Cert as an experimental feature.

Reviewer: ```suggestion
Istio provides the ability to manage settings like X-Forwarded-For and X-Forwarded-Client-Cert, which are dependent on how the gateway workloads are deployed. This is as currently an experimental feature.
```

Many users choose to deploy Istio ingress gateways in using various network topologies
(e.g. behind Cloud Load Balancers, a self-managed Load Balancer or directly expose the

> Istio ingress gateway). As such, these topologies require different ingress gateway configurations for

Reviewer: ```suggestion
Istio ingress gateway to the internet). As such, these topologies require different ingress gateway configurations for
```

> correct client attributes like IP/ports/certs to be propagated to the workload running in the cluster.

Reviewer: ```suggestion
transporting correct client attributes like IP/certs to the workloads running in the cluster.
```

Configuration of XFF and XFCC headers are configured by using by using `MeshConfig` during Istio
*installation*. To simplify configuration, create a single YAML file to pass to `istioctl`.

> {{< text yaml >}}

Author: @nrjpoddar should we mention Pod annotations?

Reviewer: yeah, let's add a section about how gateway workloads can override the default setting.

Additionally, mention that after you have edited the mesh config setting, the gateway pods need to be restarted.

Author: Done.

cat <<'EOF' > topology.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    defaultConfig:
      gatewayTopology:
EOF
{{< /text >}}

### Configuring X-Forwarded-For Headers

> Many applications, both old and new, rely on client attributes, such as X-Forward-For or X-Forwarded-Proto,

Reviewer: ```suggestion
Many applications, both old and new, rely on client attributes, such as X-Forward-For header,
```

to be forwarded by reverse proxies in a request. However, due to the variety of network
topologies Istio can be deployed in, it is required to set the number of trusted proxies deployed in front

> of the Istio gateway proxy.

Reviewer: ```suggestion
of the Istio gateway proxy so that the client address can be extracted correctly.
```

To set the number of trusted proxies add the following to your `topology.yaml` file.

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    defaultConfig:
      gatewayTopology:
        numTrustedProxies: <VALUE>
{{< /text >}}

For example, if you have a cloud based LB, a reverse proxy and then the Istio gateway proxy

> then `<VALUE>` would be 3.

Reviewer: ```suggestion
then `<VALUE>` would be 2.
```

Reviewer: Note that all the proxies in front need to parse HTTP traffic and append the X-Forwarded-For headers for every hop. If the number of entries in `X-Forwarded-For` header is less than the number of trusted hops configured, Envoy falls back to using the immediate downstream address as the trusted client address.  Please refer to [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for) for understanding how `X-Forwarded-For` headers and trusted client address is determined by Envoy.

Author: added
